include/pagmo/algorithms/gaco.hpp

```diff
@@ -0,0 +1,988 @@
+        if (m_ker > pop_size) {
+            pagmo_throw(std::invalid_argument,
+                        get_name() + " cannot work with a solution archive bigger than the population size");
+        }
+        if (n_obj != 1u) {
+            pagmo_throw(std::invalid_argument, "Multiple objectives detected in " + prob.get_name() + " instance. "
+                                                   + get_name() + " cannot deal with them");
+        }
+
+        // ---------------------------------------------------------------------------------------------------------
+
+        // No throws, all valid: we clear the logs
+        m_log.clear();
+
+        // Main ACO loop over generations:
+        for (decltype(m_gen) gen = 1u; gen <= m_gen; ++gen) {
+            // At each generation we make a copy of the population into popold
+            population popold(pop);
+
+            auto dvs = pop.get_x(); // note that pop.get_x()[n][k] goes through the different individuals of the
+            // population (index n) and the number of variables (index k) the number of
+            // variables can be easily be deduced from counting the bounds.
+            auto fit = pop.get_f(); // The following returns a vector of vectors in which objectives, equality and
+            // inequality constraints are concatenated,for each individual
+
+            // I check whether the maximum number of function evaluations or improvements has been exceeded:
+            if (m_impstop != 0 && m_n_impstop >= m_impstop) {
+                std::cout << "max number of impstop exceeded" << std::endl;
+                return pop;
+            }
+
+            if (m_evalstop != 0 && m_n_evalstop >= m_evalstop) {
+                std::cout << "max number of evalstop exceeded" << std::endl;
+                return pop;
+            }
+
+            // 1 - compute penalty functions
+
+            // I declare some useful variables:
+            vector_double penalties(pop_size);
+
+            for (decltype(pop_size) i = 0u; i < pop_size; ++i) {
+                // Penalty computation is here executed:
+                penalties[i] = penalty_computation(fit[i], pop, n_obj, n_ec, n_ic);
+            }
+
+            // 2 - update and sort solutions in the sol_archive, based on the computed penalties
+
+            // I declare a vector where the penalties are sorted:
+            vector_double sorted_penalties(penalties);
+            // This sorts the penalties from the smallest ([0]) to the biggest ([end])
+            std::sort(sorted_penalties.begin(), sorted_penalties.end(), detail::less_than_f<double>);
+            // I declare a vector where I will store the positions of the various individuals:
+            std::vector<decltype(penalties.size())> sort_list(penalties.size());
+            // We fill it with 0,1,2,3,...,K
+            std::iota(std::begin(sort_list), std::end(sort_list), 0);
```

Comment:
This usage of ``iota`` is incorrect. You need to use not ``0``, but ``decltype(penalties.size())(0)``, because otherwise ``iota`` uses signed integers for the generation of the sequence.